File: backend/app/services/redis_cache.py
from __future__ import annotations

import json
from datetime import datetime, timezone

import redis

from app.schemas.prediction import PredictionResult
from app.schemas.sensor import SensorPayload


class RedisCache:
    def __init__(self, redis_url: str, key_prefix: str, ttl_seconds: int) -> None:
        self.client = redis.Redis.from_url(redis_url, decode_responses=True)
        self.key_prefix = key_prefix
        self.ttl_seconds = ttl_seconds

    def _machine_prefix(self, machine_id: str) -> str:
        return f"{self.key_prefix}:machine:{machine_id}"

    def update_live_state(self, payload: SensorPayload, prediction: PredictionResult, mongo_record_id: str) -> dict[str, str]:
        machine_prefix = self._machine_prefix(payload.machine_id)
        live_key = f"{machine_prefix}:live"
        sensor_key = f"{machine_prefix}:sensor"
        prediction_key = f"{machine_prefix}:prediction"

        sensor_payload = payload.model_dump(mode="json")
        prediction_payload = prediction.model_dump(mode="json")
        state = {
            "machine_id": payload.machine_id,
            "sensor": sensor_payload,
            "prediction": prediction_payload,
            "mongo_record_id": mongo_record_id,
            "updated_at": datetime.now(timezone.utc).isoformat(),
        }

        pipeline = self.client.pipeline()
        pipeline.setex(sensor_key, self.ttl_seconds, json.dumps(sensor_payload))
        pipeline.setex(prediction_key, self.ttl_seconds, json.dumps(prediction_payload))
        pipeline.setex(live_key, self.ttl_seconds, json.dumps(state))
        pipeline.execute()

        return {
            "live": live_key,
            "sensor": sensor_key,
            "prediction": prediction_key,
        }

File: node-backend/src/services/redisCache.js
const { createClient } = require("redis");

class RedisCache {
  constructor(url, keyPrefix, ttlSeconds) {
    this.url = url;
    this.keyPrefix = keyPrefix;
    this.ttlSeconds = ttlSeconds;
    this.client = null;
  }

  async connect() {
    if (this.client) {
      return this.client;
    }

    this.client = createClient({ url: this.url });
    this.client.on("error", (error) => {
      console.error("Redis error:", error);
    });
    await this.client.connect();
    return this.client;
  }

  buildKey(machineId, suffix) {
    return `${this.keyPrefix}:live:${machineId}:${suffix}`;
  }

  async storeLiveReading(machineId, sensorPayload, prediction) {
    const client = await this.connect();
    const keys = {
      snapshot: this.buildKey(machineId, "snapshot"),
      sensor: this.buildKey(machineId, "sensor"),
      prediction: this.buildKey(machineId, "prediction")
    };

    const snapshot = {
      machine_id: machineId,
      sensor: sensorPayload,
      prediction,
      updated_at: new Date().toISOString()
    };

    await client
      .multi()
      .set(keys.snapshot, JSON.stringify(snapshot), { EX: this.ttlSeconds })
      .set(keys.sensor, JSON.stringify(sensorPayload), { EX: this.ttlSeconds })
      .set(keys.prediction, JSON.stringify(prediction), { EX: this.ttlSeconds })
      .exec();

    return keys;
  }

  async close() {
    if (!this.client) {
      return;
    }

    await this.client.quit();
    this.client = null;
  }
}

module.exports = {
  RedisCache
};